### ISSUE TITLE
Potential RotaryEmbedding error?

### ISSUE THREAD

**Issue by chenmoneygithub** [keras-team/keras-nlp#1416]
Hi team,

I was checking the implementation of `RotaryEmbedding` layer, and was a bit confused at the following computation:

```
def _apply_rotary_pos_emb(self, tensor, cos_emb, sin_emb):
    x1, x2 = ops.split(tensor, 2, axis=self.feature_axis)
    half_rot_tensor = ops.concatenate((-x2, x1), axis=self.feature_axis)
    return (tensor * cos_emb) + (half_rot_tensor * sin_emb)
```

Here `ops.split(tensor, 2, axis=self.feature_axis)` should be splitting the tensor by the middle, i.e., `[0, 1, 2, 3] => [0, 1], [2, 3]`, but from the paper it should be `[0, 2], [1, 3]` 
<img width="624" alt="Screenshot 2024-01-28 at 3 13 23 PM" src="https://github.com/keras-team/keras-nlp/assets/22925031/810bf4f0-d95f-42a6-8cd5-ac4932c3200e">
. Can I get some clarification? thanks!

**Comment by chenmoneygithub** [comment]
@mattdangerw Since LLaMa exists in KerasNLP now, I assume the routine input/output checks have been done, so I am more confused about this implementation. 

here is how another repo do the half rotation:
```
def rotate_half(x):
    x = rearrange(x, '... (d r) -> ... d r', r = 2)
    x1, x2 = x.unbind(dim = -1)
    x = torch.stack((-x2, x1), dim = -1)
    return rearrange(x, '... d r -> ... (d r)')
```

**Comment by mattdangerw** [comment]
@chenmoneygithub hi!!

Will take a look soon! cc @shivance in case he has thoughts (he wrote this originally)

**Comment by tirthasheshpatel** [comment]
Hi Chen,

> Here `ops.split(tensor, 2, axis=self.feature_axis)` should be splitting the tensor by the middle, i.e., `[0, 1, 2, 3] => [0, 1], [2, 3]`, but from the paper it should be `[0, 2], [1, 3]`

This is the fundamental difference beween the original PyTorch implementation [[1]] of the rotary embeddings and HuggingFace's implementation [[2]]. They should still compute the same thing if the weights are permuted correctly [[3]].

LLaMA ported the weigghts over from Huggingface directly so the outputs of the KerasNLP model and the Huggingface model matched [[4]]. On the other hand, Mistral was ported from the original PyTorch implementation and I had to add [this thin layer](https://github.com/keras-team/keras-nlp/blob/a4088f2a50007820ea0bca9cbc37129fad086f56/keras_nlp/models/mistral/mistral_attention.py#L156) around the rotary embeddings layer in KerasNLP for the outputs to match.

[1]: https://github.com/facebookresearch/llama/blob/ef351e9cd9496c579bf9f2bb036ef11bdc5ca3d2/llama/model.py#L80-L104
[2]: https://github.com/huggingface/transformers/blob/06901162b53918c302306b65b90c6a8ec99e0d0b/src/transformers/models/llama/modeling_llama.py#L123-L156
[3]: https://github.com/huggingface/transformers/blob/06901162b53918c302306b65b90c6a8ec99e0d0b/src/transformers/models/mistral/convert_mistral_weights_to_hf.py#L2
[4]: https://github.com/keras-team/keras-nlp/blob/a4088f2a50007820ea0bca9cbc37129fad086f56/tools/checkpoint_conversion/convert_llama_checkpoints.py
